src/ctap/storage.rs

```diff
@@ -458,6 +462,70 @@ impl PersistentStore {
         )?)
     }
 
+    /// Reads the byte vector stored as the serialized large blobs array.
+    ///
+    /// If more data is requested than stored, return as many bytes as possible.
+    pub fn get_large_blob_array(
+        &self,
+        mut byte_count: usize,
+        mut offset: usize,
+    ) -> Result<Vec<u8>, Ctap2StatusCode> {
+        if self.store.find(key::LARGE_BLOB_SHARDS.start)?.is_none() {
+            return Ok(vec![
+                0x80, 0x76, 0xbe, 0x8b, 0x52, 0x8d, 0x00, 0x75, 0xf7, 0xaa, 0xe9, 0x8d, 0x6f, 0xa5,
+                0x7a, 0x6d, 0x3c,
+            ]);
+        }
+        let mut output = Vec::with_capacity(byte_count);
+        while byte_count > 0 {
+            let shard = offset / SHARD_SIZE;
+            let shard_offset = offset % SHARD_SIZE;
+            let shard_length = cmp::min(SHARD_SIZE - shard_offset, byte_count);
+
+            let shard_key = key::LARGE_BLOB_SHARDS.start + shard;
+            if !key::LARGE_BLOB_SHARDS.contains(&shard_key) {
+                // This request should have been caught at application level.
+                return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+            }
+            let shard_entry = self.store.find(shard_key)?.unwrap_or_default();
+            if shard_entry.len() < shard_offset + shard_length {
+                output.extend(&shard_entry[..]);
```

Comment:
This looks wrong. Probably `&shard_entry[shard_offset..]` with an additional check that `shard_offset <= shard_entry.len()`.

Reply:
The specification asks to return all remaining bytes, if more are requested than exist. This is what this line is about. Maybe the confusion here and below is from `shard_length` not being the length of the shard in storage, but the length of what we want to read from this shard. Maybe another variable name works better for you?
(Also, if it was wrong, it should have been caught by unit tests.)

Reply:
The issue I have in mind is:
- one shard of 10 bytes
- asking to read from offset 1 (i.e. skipping first byte)
- asking to read 20 bytes

My understanding is that we don't skip the first byte. So I guess when you say "return all remaining bytes" it's not just the ones after the offset, but the whole content. Then there would still be a problem with the following case:

- 2 shards A and B
- offset is A.len + 1
- length is A.len + B.len + 1

My understanding is that we return only B and not AB.

Regarding unit tests, how could they catch all bugs? Unit tests are just basic checks, they can't guarantee anything about the code (as long as there are a few ifs or a loop).

Reply:
Oh true, I thought I had this covered. Fixed.

Reply:
Also, unit test added for coverage.